src/Microsoft.OData.Client/Materialization/FeedAndEntryMaterializerAdapter.cs

```diff
@@ -296,6 +323,87 @@ private ODataResourceSet ReadFeedCore(bool lazy)
+        /// Reads an <see cref="ODataDeltaResourceSet"/>.
+        /// </summary>
+        /// <returns>The <see cref="MaterializerDeltaFeed"/> of the read <see cref="ODataDeltaResourceSet"/>.</returns>
+        private MaterializerDeltaFeed ReadDeltaFeedCore()
+        {
+            this.ExpectState(ODataReaderState.DeltaResourceSetStart);
+
+            ODataDeltaResourceSet result = (ODataDeltaResourceSet)this.reader.Item;
+            MaterializerDeltaFeed feed = null;
+
+            if (result == null)
+            {
+                return null;
+            }
+
+            feed = MaterializerDeltaFeed.CreateDeltaFeed(
+                    result,
+                    new List<IMaterializerState>(),
+                    this.materializerContext);
+
+            if (this.items?.Count > 0)
+            {
+                this.items.Push(feed);
+            }
+            else
+            {
+                IMaterializerState topLevelItem = feed;
+                this.items.Push(topLevelItem);
+            }
+
+            do
+            {
+                this.AssertRead();
+
+                switch (this.reader.State)
+                {
+                    case ODataReaderState.ResourceStart:
+                        MaterializerEntry entry = this.ReadEntryCore();
+                        feed.AddEntry(entry);
+                        break;
+                    case ODataReaderState.DeletedResourceStart:
+                        MaterializerDeletedEntry deletedMaterializerEntry = this.ReadDeletedResource();
+                        feed.AddEntry(deletedMaterializerEntry);
+                        break;
+                    case ODataReaderState.DeletedResourceEnd:
+                        break;
+                    case ODataReaderState.ResourceEnd:
+                    case ODataReaderState.DeltaResourceSetEnd:
+                        this.items?.Pop();
+                        break;
+                    default:
+                        throw DSClient.Error.InternalError(InternalError.UnexpectedReadState);
+                }
+            }
+            while (this.reader.State != ODataReaderState.DeltaResourceSetEnd);
+
+            return feed;
+        }
+
+        /// <summary>
+        /// Adds an <see cref="IMaterializerState"/> entry to a parent entry.
+        /// </summary>
+        /// <param name="materializer">The <see cref="IMaterializerState"/> entry being added to the parent item.</param>
+        private void AddResourceToParent(IMaterializerState materializer)
+        {
+            IMaterializerState topItem = this.items.Peek();
+
+            if (topItem != null)
+            {
+                if (topItem is MaterializerDeltaFeed delta)
+                {
+                    delta.AddEntry(materializer);
+                }
+                else if (topItem is MaterializerEntry entry)
+                {
+                    entry.AddNestedItem(materializer);
+                }
+            }

```

Comment:
I don't think you need the outer `if` block
```suggestion
                if (topItem is MaterializerDeltaFeed delta)
                {
                    delta.AddEntry(materializer);
                }
                else if (topItem is MaterializerEntry entry)
                {
                    entry.AddNestedItem(materializer);
                }
```